Mathlib/Topology/UniformSpace/Completion.lean

```diff
@@ -0,0 +1,728 @@
+/-
+Copyright (c) 2018 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot, Johannes Hölzl
+
+! This file was ported from Lean 3 source module topology.uniform_space.completion
+! leanprover-community/mathlib commit dc6c365e751e34d100e80fe6e314c3c3e0fd2988
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Topology.UniformSpace.AbstractCompletion
+
+/-!
+# Hausdorff completions of uniform spaces
+
+The goal is to construct a left-adjoint to the inclusion of complete Hausdorff uniform spaces
+into all uniform spaces. Any uniform space `α` gets a completion `Completion α` and a morphism
+(ie. uniformly continuous map) `coe : α → Completion α` which solves the universal
+mapping problem of factorizing morphisms from `α` to any complete Hausdorff uniform space `β`.
+It means any uniformly continuous `f : α → β` gives rise to a unique morphism
+`Completion.extension f : Completion α → β` such that `f = Completion.extension f ∘ coe`.
+Actually `Completion.extension f` is defined for all maps from `α` to `β` but it has the desired
+properties only if `f` is uniformly continuous.
+
+Beware that `coe` is not injective if `α` is not Hausdorff. But its image is always
+dense. The adjoint functor acting on morphisms is then constructed by the usual abstract nonsense.
+For every uniform spaces `α` and `β`, it turns `f : α → β` into a morphism
+  `Completion.map f : Completion α → Completion β`
+such that
+  `coe ∘ f = (Completion.map f) ∘ coe`
+provided `f` is uniformly continuous. This construction is compatible with composition.
+
+In this file we introduce the following concepts:
+
+* `CauchyCat α` the uniform completion of the uniform space `α` (using Cauchy filters).
+  These are not minimal filters.
+
+* `Completion α := quotient (separation_setoid (CauchyCat α))` the Hausdorff completion.
+
+## References
+
+This formalization is mostly based on
+  N. Bourbaki: General Topology
+  I. M. James: Topologies and Uniformities
+From a slightly different perspective in order to reuse material in topology.uniform_space.basic.
+-/
+
+
+noncomputable section
+
+open Filter Set
+
+universe u v w x
+
+open Uniformity Classical Topology Filter
+
+/-- Space of Cauchy filters
+
+This is essentially the completion of a uniform space. The embeddings are the neighbourhood filters.
+This space is not minimal, the separated uniform space (i.e. quotiented on the intersection of all
+entourages) is necessary for this.
+-/
+def CauchyCat (α : Type u) [UniformSpace α] : Type u :=
+  { f : Filter α // Cauchy f }
+set_option linter.uppercaseLean3 false in
+#align Cauchy CauchyCat
```

Comment:
```suggestion
def Cauchy (α : Type u) [UniformSpace α] : Type u :=
  { f : Filter α // Cauchy f }
set_option linter.uppercaseLean3 false in
#align Cauchy Cauchy
```

Reply:
Right, that Cat in the name made no sense. But now we have an issue : Cauchy is already taken. We need a new name.